### PR TITLE
Fix problem with COMPOUND Conditionals

### PR DESCRIPTION
**Pull request by tj-sun**
Recursively factorise then and else branches of Conditional nodes.

Questions:
- Do we need to look into the condition node of Conditionals (and Comparison nodes as a result)
- The checks for Failure are duplicated, maybe need to move that code outside of main optimisation routine.

### CODE REVIEW

File: tsfc/coffee_mode.py [hunk 1]
```diff
@@ -48,6 +50,34 @@ def predicate(index):
...
+    raise AssertionError("cannot handle type %s" % type(node))
+
+
+_handle_conditional.register(Node)(reuse_if_untouched)
+
+
+@_handle_conditional.register(Conditional)
+def _handle_conditional_conditional(node, self):
+    condition, then, else_ = node.children
+    then, else_ = optimise_expressions([then, else_], self.argument_multiindices)
```
Comment (wence-) [inline tsfc/coffee_mode.py]: The condition could be arbitrarily complex, should we descend there too?

**Comment by miklos1** [comment]
I am not convinced that this will produce a good optimisation.

**Comment by miklos1** [comment]
We shall check out what UFLACS does in this case.

**Comment by miklos1** [comment]
I decline this.

*Reason:* this is a very questionable "fix", and I am now convinced that we shall use the following substitution in some way:

    Conditional(a, b, c) => Conditional(a, 1, 0)*b + Conditional(a, 0, 1)*c

**Pull request closed without merging**